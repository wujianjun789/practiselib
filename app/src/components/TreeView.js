/**
 * Created by a on 2017/7/3.
 */
import React, { Component } from 'react'
import {Link} from 'react-router';

import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {onToggle} from '../common/actions/treeView'

import {getLanguage} from '../util/index'
export class TreeView extends Component{
    constructor(props){
        super(props)
        this.state = {
            language: getLanguage()
        },

        this.renderTree = this.renderTree.bind(this);
        this.onToggle = this.onToggle.bind(this);
    }

    componentWillMount(){
    }

    onToggle(node){
        const {actions} = this.props
        actions && actions.onToggle(node)
        this.props.onToggle && this.props.onToggle(node);

    }

    renderTree(datalist, index, toggled){
        if(!datalist){
            return null;
        }
        let curIndex = index;
        let nextIndex = index + 1;
        let style = {"height":index>1 ? (toggled ? datalist.length*40+'px':'0'):'auto'};
        return <ul className={"tree-"+curIndex} style={style}>
            {
                datalist.map((node, index)=> {
                    let count = this.state.language=='zh'?6:10;
                    let value = node.name.slice(0, count)+(node.name.length>count?'...':'');
                    return <li key={index} className={'node '+(node.active ? 'active':'')}>
                        <Link to={node.link}>
                        <div onClick={()=>this.onToggle(node)} title={node.name}><span className={'glyphicon '+(curIndex > 1 ? (node.class+(node.active ? '_hover':''))
                        : (node.toggled ? 'glyphicon-triangle-bottom':'glyphicon-triangle-right'))}></span>
                            {value}</div></Link>
                        {node.children && this.renderTree(node.children, nextIndex, node.toggled)}
                    </li>
                })
            }
        </ul>
    }

    render(){
        const {datalist} = this.props;
        return <div className="tree-list">
            {
                this.renderTree(datalist, 1)
            }
        </div>
    }
}

function mapStateToProps(state) {
    return {
        datalist: state.treeView.datalist
    }
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators({
            onToggle: onToggle
        }, dispatch)
    }
}
export default connect(
    mapStateToProps,
    mapDispatchToProps
)(TreeView);