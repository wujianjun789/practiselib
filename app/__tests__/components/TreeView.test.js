/**
 * Created by a on 2017/7/3.
 */
import React from 'react';
import {shallow} from 'enzyme'
import renderer from 'react-test-renderer'
import TreeView from '../../src/components/TreeView';

const TreeData=[
    {
        "name":"资产管理",
        "toggled": true,
        "active": true,
        "children": [
            {
                "name":"灯集中控制器灯集中控",
                "class":"glyphicon-list-alt",
                "active":false
            },
            {
                "name":"LED灯",
                "class":"glyphicon-lamp",
                "active":false
            }
        ]
    },
    {
        "name":"资产统计",
        "toggled": false,
        "active": true
    }
]
test('TreeView renders', ()=>{

    const component = renderer.create(
        <TreeView datalist={TreeData} />
    )

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
})

test('TreeView div click', ()=>{
    const component = shallow(<TreeView datalist={TreeData}/>)

    expect(component.find('.node').length).toEqual(4);

    component.find('.node').at(1).find('div').simulate('click');
    expect(component.find('.node').at(1).props().className).toEqual('node active');
})
