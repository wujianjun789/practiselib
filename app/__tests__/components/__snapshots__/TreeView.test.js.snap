exports[`test TreeView renders 1`] = `
<div
  className="tree-list undefined">
  <ul
    className="tree-1"
    style={
      Object {
        "height": "auto",
      }
    }>
    <li
      className="node active">
      <a
        onClick={[Function]}
        style={Object {}}>
        <div
          onClick={[Function]}
          title="资产管理">
          <span
            className="glyphicon glyphicon-triangle-bottom toggled" />
          资产管理
        </div>
      </a>
      <ul
        className="tree-2"
        style={
          Object {
            "height": "80px",
          }
        }>
        <li
          className="node ">
          <a
            onClick={[Function]}
            style={Object {}}>
            <div
              onClick={[Function]}
              title="灯集中控制器灯集中控">
              <span
                className="icon glyphicon-list-alt" />
              <span>
                灯集中控制器...
              </span>
            </div>
          </a>
        </li>
        <li
          className="node ">
          <a
            onClick={[Function]}
            style={Object {}}>
            <div
              onClick={[Function]}
              title="LED灯">
              <span
                className="icon glyphicon-lamp" />
              <span>
                LED灯
              </span>
            </div>
          </a>
        </li>
      </ul>
    </li>
    <li
      className="node active">
      <a
        onClick={[Function]}
        style={Object {}}>
        <div
          onClick={[Function]}
          title="资产统计">
          <span
            className="icon undefined" />
          <span>
            资产统计
          </span>
        </div>
      </a>
    </li>
  </ul>
</div>
`;
